lib/utils.js

```diff
@@ -0,0 +1,82 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2020 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+/**
+ * 
+ * @param {any} object Object to test
+ * @returns {boolean} true if it's a plain object, false otherwise
+ */
+function isPlainObject (object) {
+  return object !== null && typeof object === 'object' && !Array.isArray(object);
+}
+
+/**
+ * 
+ * @param {Object} object Plain object to map
+ * @param {(key: string, value: any, index: number) => {key: string, value: any}} fn Map function returning an object containing what must be updated (key/value)
+ * @returns {Object} Mapped object
+ */
+const objectRecursiveMap = (object, fn) => {
+  return Object.fromEntries(
+    Object.entries(object).map(
+      ([key, value], index) => {
+        let { key: newKey = key, value: newValue = value } = fn(key, value, index);
+
+        if (Array.isArray(newValue)) {
+          newValue = newValue.map(item => 
+            typeof item === 'object' ? objectRecursiveMap(item, fn) : item);
+        }
+        else if (typeof newValue === 'object') {
```

Comment:
You should use `isPlainObject` here instead. 

```
> typeof new Map()
'object'
> typeof new Date()
'object'
```

Reply:
I'm aware... Unfortunately, `isPlainObject` has its limits too... This would still be considered as a plain object...
`isPlainObject` returns only false when `null`, if its an array or if not an object.

I can indeed check for `null` (and it can't be an array because of preceding `if`) but if we want more precision, we will need to modify `isPlainObject`. What do you think?

Reply:
I think we should modify `isPlainObject` then. 
Maybe there is something to do with constructor names but it will need a little bit of investigation
```
> m = new Map()
Map {}
> m.constructor.name
'Map'
> o = {}
{}
> o.constructor.name
'Object'
> a = []
[]
> a.constructor.name
'Array'
```

Reply:
isPlainObject has just been fixed in kuzzleio/kuzzle#2070